rs_inverse/scripts/env_tester.py

```diff
@@ -106,225 +256,47 @@ def talker(motors_target_pos):
     # stand up
     time.sleep(2.1)
-    talker(stand_up)
-    rate.sleep()
-    time.sleep(3.1)
-
-    rospy.loginfo_once("STARTED SPOT TEST ENV")
+    spot_control.talker(stand_up)
+    spot_control.rate.sleep()
+    time.sleep(1.1)
+    # ---
     t = 0
-    global xd, yd, zd, rolld, pitchd, yawd, StepLength, LateralFraction, YawRate, StepVelocity, ClearanceHeight, \
-        PenetrationDepth, SwingPeriod
-    global x_inst, y_inst, z_inst, roll_inst, pitch_inst, yaw_inst, vx_inst, vy_inst, vz_inst, wz_inst
-    global front_left_lower_leg_contact, front_right_lower_leg_contact, rear_left_lower_leg_contact, rear_right_lower_leg_contact
-
-    listener()
-
-    y_target = 1.5
-    rospy.loginfo_once("y_target_init=" + str(y_target))
-
-    pid = PID(0.1, 0.01, 0.05)
-    pid.sample_time = 0.01
-    pid.output_limits = (-1, 1)
-    pid.setpoint = y_target
-    pid.proportional_on_measurement = True
+    spot_control.listener()
+    rospy.loginfo_once("STARTED SPOT TEST ENV")
 
     while t < (int(max_timesteps)):
-        pos = np.array([xd, yd, zd])
-        orn = np.array([rolld, pitchd, yawd])
-
-        # Update Swing Period
-        bzg.Tswing = SwingPeriod
-
-        # rospy.loginfo_throttle(0.005, "contact front left = " + str(front_left_lower_leg_contact))
-        # rospy.loginfo_throttle(0.005, "contact front left = " + str(front_right_lower_leg_contact))
-        # rospy.loginfo_throttle(0.005, "contact rear left = " + str(rear_left_lower_leg_contact))
-        # rospy.loginfo_throttle(0.005, "contact rear right = " + str(rear_right_lower_leg_contact))
-        contacts = [front_left_lower_leg_contact, front_right_lower_leg_contact, rear_left_lower_leg_contact,
-                    rear_right_lower_leg_contact]
-
-        residual = y_target - y_inst
-        if abs(residual) < 1.0:
-            p_yaw = 0
-        elif 0.5 > abs(residual) >= 0.1:
-            p_yaw = 0.5
-        else:
-            p_yaw = 0.1
-
-        rospy.loginfo_throttle(1, "residual=" + str(y_target - y_inst))
-
-        pid.Kp = p_yaw
-        YawRate_d = pid(y_inst)
-
-        # Get Desired Foot Poses
-        T_bf = bzg.GenerateTrajectory(StepLength, LateralFraction, YawRate_d,
-                                      StepVelocity, T_bf0, T_bf,
-                                      ClearanceHeight, PenetrationDepth,
-                                      contacts)
-        joint_angles = spot.IK(orn, pos, T_bf)
+        start_time = time.time()
 
+        spot_control.spot_inverse_control()
 
         t += 1
-        # print(joint_angles)
-        talker(joint_angles)
-        rate.sleep()
+        # elapsed time == sample time
+        elapsed_time = time.time() - start_time
+        if elapsed_time < spot_control.time_step:
+            time.sleep(spot_control.time_step - elapsed_time)
+
 
+def myhook():
+    # Execute on shutdown
+    print("FINISHED SPOT TEST ENV")
 
 if __name__ == '__main__':
     try:
         main()
+        rospy.on_shutdown(myhook)
```

Comment:
In a single-threaded script, you don't need this hook as it is set only after the main function is fully executed.
It makes set to set on_shutdown hook before calling the main function. The hook is used to clear some variables, send notifications, etc.

Reply:
Got it